### PR TITLE
ARROW-14048: [C++][Gandiva] Cache only object code in memory instead of entire module

### PR DESCRIPTION
**Pull request by augustoasilva**
Implement Gandiva to cache object code instead the entire llvm module, improving the memory consumption.

To compare the optmization, the following test was written to verify the memory allocation used by the gandiva cache previous to this, and after this:

```cpp
TEST_F(TestProjector, TestObjCache) {
  // schema for input fields
  auto field0 = field("f0", int32());
  auto field1 = field("f2", int32());
  auto field2 = field("f0", arrow::float64());
  auto field3 = field("f2", arrow::float64());
  auto field4 = field("f0", arrow::utf8());

  auto schema1 = arrow::schema({field0, field1});
  auto schema2 = arrow::schema({field2, field3});
  auto schema3 = arrow::schema({field4});

  // output fields
  auto field_sum = field("add", int32());
  auto field_sub = field("subtract", int32());
  auto field_mul = field("multiply", int32());
  auto field_div = field("divide", int32());
  auto field_eq = field("equal", arrow::boolean());
  auto field_lt = field("less_than", arrow::boolean());
  auto field_logb = arrow::field("logb", arrow::float64());
  auto field_power = arrow::field("power", arrow::float64());
  auto res_float8 = field("res_float8", arrow::float64());
  auto res_int4 = field("castBIGINT", arrow::int32());

  // Build expressions for schema1
  auto sum_expr = TreeExprBuilder::MakeExpression("add", {field0, field1}, field_sum);
  auto sub_expr =
      TreeExprBuilder::MakeExpression("subtract", {field0, field1}, field_sub);
  auto mul_expr =
      TreeExprBuilder::MakeExpression("multiply", {field0, field1}, field_mul);
  auto div_expr = TreeExprBuilder::MakeExpression("divide", {field0, field1}, field_div);
  auto eq_expr = TreeExprBuilder::MakeExpression("equal", {field0, field1}, field_eq);
  auto lt_expr = TreeExprBuilder::MakeExpression("less_than", {field0, field1}, field_lt);

  // Build expressions for schema2
  auto logb_expr = TreeExprBuilder::MakeExpression("log", {field2, field3}, field_logb);
  auto power_expr =
      TreeExprBuilder::MakeExpression("power", {field2, field3}, field_power);

  // Build expressions for schema3
  auto cast_expr_float8 =
      TreeExprBuilder::MakeExpression("castFLOAT8", {field4}, res_float8);
  auto cast_expr_int8 = TreeExprBuilder::MakeExpression("castINT", {field4}, res_int4);

  auto configuration = TestConfiguration();

  // Uses field0, field1 and schema1
  ExpressionVector exprVec1 = {sum_expr};
  ExpressionVector exprVec2 = {sub_expr};
  ExpressionVector exprVec3 = {mul_expr};
  ExpressionVector exprVec4 = {div_expr};
  ExpressionVector exprVec5 = {eq_expr};
  ExpressionVector exprVec6 = {lt_expr};

  // Uses field2, field3 and schema2
  ExpressionVector exprVec7 = {logb_expr};
  ExpressionVector exprVec8 = {power_expr};

  // Uses field4 and schema3
  ExpressionVector exprVec9 = {cast_expr_float8};
  ExpressionVector exprVec10 = {cast_expr_int8};

  unsigned int microsecond = 1000000;
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 1st projector
  std::shared_ptr<Projector> projector1;
  auto status = Projector::Make(schema1, exprVec1, configuration, &projector1);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // delete projector1 // -> gives the following error:
  // error: type ‘class std::shared_ptr<gandiva::Projector>’ argument given to ‘delete’, expected pointer

  //delete projector1.get(); // -> gives the following error:
  // double free() call. Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

  // If this line is looged, thant the shared_ptr will be the only ptr holding the projector
  // so reseting it will explicit delete it.
  if(projector1.unique()){
    ARROW_LOG(INFO) << "ONLY PTR TO THE PROJECTOR INSTANCE.";
  }
  // safer way to delete the projector as the deleter param of shared_ptr is the delete func by default.
  // ref: https://en.cppreference.com/w/cpp/memory/shared_ptr/reset
  projector1.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 2nd projector
  std::shared_ptr<Projector> projector2;
  status = Projector::Make(schema1, exprVec2, configuration, &projector2);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector2.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 3rd projector
  std::shared_ptr<Projector> projector3;
  status = Projector::Make(schema1, exprVec3, configuration, &projector3);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector3.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 4th projector
  std::shared_ptr<Projector> projector4;
  status = Projector::Make(schema1, exprVec4, configuration, &projector4);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector4.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 5th projector
  std::shared_ptr<Projector> projector5;
  status = Projector::Make(schema1, exprVec5, configuration, &projector5);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector5.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 6th projector
  std::shared_ptr<Projector> projector6;
  status = Projector::Make(schema1, exprVec6, configuration, &projector6);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector6.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 7th projector
  std::shared_ptr<Projector> projector7;
  status = Projector::Make(schema2, exprVec7, configuration, &projector7);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector7.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 8th projector
  std::shared_ptr<Projector> projector8;
  status = Projector::Make(schema2, exprVec8, configuration, &projector8);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector8.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 9th projector
  std::shared_ptr<Projector> projector9;
  status = Projector::Make(schema3, exprVec9, configuration, &projector9);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector9.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second

  // 10th projector
  std::shared_ptr<Projector> projector10;
  status = Projector::Make(schema3, exprVec10, configuration, &projector10);
  ASSERT_OK(status);
  usleep(0.5 * microsecond);//sleeps for 0.5 second

  projector10.reset();

  usleep(0.5 * microsecond);//sleeps for 0.5 second
}
```

The results can be seen here:
|Projector   |Module Cache Size (Bytes)|Object Code Cache Size (Bytes)|Optmization|
|------------|-----------|----------------|-----------|
|Projector 01| 1,640,843   | 97,920          | 94.03%    |
|Projector 02| 1,559,247   | 16,135          | 98.97%    |
|Projector 03| 1,559,247   | 16,135          | 98.97%    |
|Projector 04| 1,558,080   | 16,622          | 98.93%    |
|Projector 05| 1,593,485   | 65,399          | 95.90%    |
|Projector 06| 1,544,122   | 16,231          | 98.95%    |
|Projector 07| 1,859,869   | 17,691          | 99.05%    |
|Projector 08| 1,822,195   | 16,095          | 99.12%    |
|Projector 09| 1,552,585   | 16,208          | 98.96%    |
|Projector 10| 1,551,472   | 16,367          | 98.95%    |

To track the memory was used the Heaptrack application.

### CODE REVIEW

File: cpp/src/gandiva/greedy_dual_size_cache.h [hunk 1]
```diff
@@ -146,9 +159,25 @@ class GreedyDualSizeCache {
...
+    map_.erase((*i).cache_key);
+    priority_set_.erase(i);
+  }
+
   map_type map_;
   std::set<PriorityItem> priority_set_;
   uint64_t inflation_;
   size_t capacity_;
+  size_t cache_size_ = 0;
+  llvm::SmallString<128> cache_dir_;
```
Comment (projjal) [inline cpp/src/gandiva/greedy_dual_size_cache.h:181]: looks like this is not getting used.

File: cpp/src/gandiva/projector.h [hunk 2]
```diff
@@ -138,6 +167,7 @@ class GANDIVA_EXPORT Projector {
   SchemaPtr schema_;
   FieldVector output_fields_;
   std::shared_ptr<Configuration> configuration_;
+  bool compiled_from_cache_;
```
Comment (projjal) [inline cpp/src/gandiva/projector.h:170]: nit: use a different name like built_from_cache since it is already compiled

File: cpp/src/gandiva/llvm_generator.h [hunk 3]
```diff
@@ -49,10 +51,48 @@ class GANDIVA_EXPORT LLVMGenerator {
...
+  GetCache();
+
   /// \brief Build the code for the expression trees for default mode. Each
   /// element in the vector represents an expression tree
   Status Build(const ExpressionVector& exprs, SelectionVector::Mode mode);
 
+  /// \brief Build the code for the expression trees for default mode with a LLVM
+  /// ObjectCache. Each element in the vector represents an expression tree
+  template <class KeyType>
+  Status Build(const ExpressionVector& exprs, SelectionVector::Mode mode,
```
Comment (projjal) [inline cpp/src/gandiva/llvm_generator.h:64]: Better to add the definition in the cc file.

File: cpp/src/gandiva/llvm_generator.h [hunk 4]
```diff
@@ -49,10 +51,48 @@ class GANDIVA_EXPORT LLVMGenerator {
...
+    // Start measuring code gen time
+    auto begin = std::chrono::high_resolution_clock::now();
+    for (auto& expr : exprs) {
+      auto output = annotator_.AddOutputFieldDescriptor(expr->result());
+      ARROW_RETURN_NOT_OK(Add(expr, output));
+    }
+
+    // Stop measuring time, calculate the elapsed time and pass it to object cache
+    auto end = std::chrono::high_resolution_clock::now();
+    size_t elapsed =
```
Comment (projjal) [inline cpp/src/gandiva/llvm_generator.h:77]: This is wrong. the elapse time was supposed to track the compilation time which is mainly done in FinalizeModule

**Review by projjal — CHANGES_REQUESTED** [review]

**Comment by projjal** [comment]
> Implement Gandiva to cache object code instead the entire llvm module, improving the memory consumption and LLVM time perfomance.

This should not have any effect on llvm performance only the memory footprint. Can you also include the numbers in the description.
Also the wording in the description and pr title seems vague. change it to something like "Cache only ObjectCode in memory instead of entire module"

File: cpp/src/gandiva/cache.h [hunk 5]
```diff
@@ -39,20 +39,30 @@ class Cache {
...
 
-  void PutModule(KeyType cache_key, ValueCacheObject<ValueType> valueCacheObject) {
+  void PutObjectCode(KeyType& cache_key, ValueCacheObject<ValueType> object_code) {
     mtx_.lock();
-    cache_.insert(cache_key, valueCacheObject);
+    cache_.InsertObjectCode(cache_key, object_code);
     mtx_.unlock();
   }
 
+  ::std::shared_ptr<Cache> CreateSharedCachePtr() { return Cache::create(); }
```
Comment (projjal) [inline cpp/src/gandiva/cache.h:60]: This looks not needed.

File: cpp/src/gandiva/tests/filter_test.cc [hunk 6]
```diff
@@ -29,7 +29,11 @@ using arrow::int32;
 
 class TestFilter : public ::testing::Test {
  public:
-  void SetUp() { pool_ = arrow::default_memory_pool(); }
+  void SetUp() {
+    pool_ = arrow::default_memory_pool();
+    // Setup arrow log severity threshold to debug level.
+    arrow::util::ArrowLog::StartArrowLog("", arrow::util::ArrowLogLevel::ARROW_DEBUG);
```
Comment (projjal) [inline cpp/src/gandiva/tests/filter_test.cc:35]: is this change required?

File: cpp/src/gandiva/filter.cc [hunk 7]
```diff
@@ -42,7 +43,7 @@ FilterCacheKey::FilterCacheKey(SchemaPtr schema,
   expression_as_string_ = expression.ToString();
   UpdateUniqifier(expression_as_string_);
   arrow::internal::hash_combine(result, expression_as_string_);
-  arrow::internal::hash_combine(result, configuration);
+  arrow::internal::hash_combine(result, configuration->Hash());
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/filter.cc:46]: I thought that configuration being a shared_ptr when using it directly to hash it hashed the memory location value, so as it has a Hash() function, it looked better to use the hash_code that it generates to use on the hash_combine. 

Do you think it is better to undo this?

File: cpp/src/gandiva/base_cache_key.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    hash_code_ = result_hash;
+    schema_ = key.schema();
+  }
+
+  size_t Hash() const { return hash_code_; }
+
+  std::string Type() const { return type_; }
+
+  bool operator==(const BaseCacheKey& other) const {
+    if (hash_code_ != other.hash_code_) {
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/base_cache_key.h:56]: Ok, I compared the other private variables to.

File: cpp/src/gandiva/greedy_dual_size_cache.h [hunk 9]
```diff
@@ -146,9 +159,25 @@ class GreedyDualSizeCache {
     priority_set_.erase(i);
   }
 
+  void evictObject() {
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/greedy_dual_size_cache.h:162]: Yeah, it was so and I should have removed it earlier, so I removed it now.

File: cpp/src/gandiva/greedy_dual_size_cache.h [hunk 10]
```diff
@@ -92,25 +96,26 @@ class GreedyDualSizeCache {
 
   bool contains(const Key& key) { return map_.find(key) != map_.end(); }
 
-  void insert(const Key& key, const ValueCacheObject<Value>& value) {
+  void InsertObjectCode(const Key& key, const ValueCacheObject<Value>& value) {
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/greedy_dual_size_cache.h:99]: Ok, fixed it.

File: cpp/src/gandiva/cache.h (hunk 5 above)
Comment (augustoasilva) [reply cpp/src/gandiva/cache.h:60]: Removed it.

File: cpp/src/gandiva/tests/filter_test.cc (hunk 6 above)
Comment (augustoasilva) [reply cpp/src/gandiva/tests/filter_test.cc:35]: Not required, just useful to debug the cache system using the tests' files. Do you want me to remove it?

File: cpp/src/gandiva/greedy_dual_size_cache.h (hunk 1 above)
Comment (augustoasilva) [reply cpp/src/gandiva/greedy_dual_size_cache.h:181]: Removed it.

File: cpp/src/gandiva/llvm_generator.cc [hunk 11]
```diff
@@ -50,6 +51,23 @@ Status LLVMGenerator::Make(std::shared_ptr<Configuration> config,
...
 
+std::shared_ptr<Cache<BaseCacheKey, std::shared_ptr<llvm::MemoryBuffer>>>
+LLVMGenerator::GetCache() {
+  static Cache<BaseCacheKey, std::shared_ptr<llvm::MemoryBuffer>> cache;
+  //  static std::unique_ptr<Cache<BaseCacheKey, std::shared_ptr<llvm::MemoryBuffer>>>
+  //      cache_unique = std::make_unique<Cache<BaseCacheKey,
+  //      std::shared_ptr<llvm::MemoryBuffer>>>();
+
+  //  static std::shared_ptr<Cache<BaseCacheKey, std::shared_ptr<llvm::MemoryBuffer>>>
+  //      shared_cache = std::move(cache_unique);
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/llvm_generator.cc:62]: Removed it

File: cpp/src/gandiva/llvm_generator.cc [hunk 12]
```diff
@@ -50,6 +51,23 @@ Status LLVMGenerator::Make(std::shared_ptr<Configuration> config,
   return Status::OK();
 }
 
+std::shared_ptr<Cache<BaseCacheKey, std::shared_ptr<llvm::MemoryBuffer>>>
+LLVMGenerator::GetCache() {
+  static Cache<BaseCacheKey, std::shared_ptr<llvm::MemoryBuffer>> cache;
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/llvm_generator.cc:56]: Removed it.

File: cpp/src/gandiva/projector.h (hunk 2 above)
Comment (augustoasilva) [reply cpp/src/gandiva/projector.h:170]: built_from_cache sounds better, so changed to it.

File: cpp/src/gandiva/llvm_generator.h (hunk 3 above)
Comment (augustoasilva) [reply cpp/src/gandiva/llvm_generator.h:64]: I know, but because this uses template, if it is on the cc file this function gets an undefined reference error for it, as just this function is using template.

File: cpp/src/gandiva/filter.cc [hunk 13]
```diff
@@ -120,17 +138,27 @@ Status Filter::Make(SchemaPtr schema, ConditionPtr condition,
   ARROW_RETURN_NOT_OK(expr_validator.Validate(condition));
 
   // Start measuring build time
-  auto begin = std::chrono::high_resolution_clock::now();
-  ARROW_RETURN_NOT_OK(llvm_gen->Build({condition}, SelectionVector::Mode::MODE_NONE));
+  //  auto begin = std::chrono::high_resolution_clock::now();
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/filter.cc:141]: Removed it

File: cpp/src/gandiva/engine.cc [hunk 14]
```diff
@@ -167,7 +167,6 @@ Status Engine::Make(const std::shared_ptr<Configuration>& conf,
     return Status::CodeGenError("Could not instantiate llvm::ExecutionEngine: ",
                                 builder_error);
   }
-
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/engine.cc:170]: Removed it

File: cpp/src/gandiva/llvm_generator.h (hunk 4 above)
Comment (augustoasilva) [reply cpp/src/gandiva/llvm_generator.h:77]: I know, but when the FinalizeModule is called, the LLVM will access the GandivaObjectCache file to execute the logic for the object code manipulation (to cache it or to get it from cache) so I need to set GandivaObjectCache before the FinalizeModule execution, and before the code sets the GandivaObjectCache it needs to know the elapsed time so the greedy dual cache logic works.

Do you have any suggestions how to proceed with that case?

File: cpp/src/gandiva/base_cache_key.h [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
```
Comment (projjal) [inline cpp/src/gandiva/base_cache_key.h:56]: There can still be collision on different expressions

File: cpp/src/gandiva/engine.h [hunk 16]
```diff
@@ -54,6 +54,13 @@ class GANDIVA_EXPORT Engine {
     functions_to_compile_.push_back(fname);
   }
 
+  /// Set BaseObjectCache.
+  template <class KeyType>
+  Status SetLLVMObjectCache(GandivaObjectCache<KeyType>& object_cache) {
+    execution_engine_->setObjectCache(&object_cache);
+    return Status::OK();
```
Comment (projjal) [inline cpp/src/gandiva/engine.h:61]: why did you remove the error detection code?

Comment (augustoasilva) [reply cpp/src/gandiva/engine.h:61]: It was failling on appveyor with the message: `error C2039: 'hasError': is not a member of 'llvm::ExecutionEngine'`

File: cpp/src/gandiva/base_cache_key.h (hunk 15 above)
Comment (augustoasilva) [reply cpp/src/gandiva/base_cache_key.h:56]: Ok, I will also check the expressions.

File: cpp/src/gandiva/llvm_generator.h (hunk 4 above)
Comment (projjal) [reply cpp/src/gandiva/llvm_generator.h:77]: You can track the time inside the gandivaObjectCache; start timer in constructor and stop it in notifyObject method?
Also, its best not to knowingly proceed with incorrect solution if you are stuck somewhere.

File: cpp/src/gandiva/llvm_generator.h (hunk 3 above)
Comment (projjal) [reply cpp/src/gandiva/llvm_generator.h:64]: Ok. It looks like there is code duplication with the other build. Should we remove the other build method if it is not used?

File: cpp/src/gandiva/tests/projector_test.cc [hunk 17]
```diff
@@ -196,21 +202,23 @@ TEST_F(TestProjector, TestProjectCacheDecimalCast) {
...
   ASSERT_OK(Projector::Make(schema, {expr0}, TestConfiguration(), &projector0));
+  EXPECT_FALSE(projector0->GetBuiltFromCache());
 
   // if the output scale is different, the cache can't be used.
   auto res_31_14 = field("result", arrow::decimal(31, 14));
   auto expr1 = TreeExprBuilder::MakeExpression("castDECIMAL", {field_float64}, res_31_14);
   std::shared_ptr<Projector> projector1;
   ASSERT_OK(Projector::Make(schema, {expr1}, TestConfiguration(), &projector1));
-  EXPECT_NE(projector0.get(), projector1.get());
+  // EXPECT_NE(projector0.get(), projector1.get()); -> old expect.
```
Comment (projjal) [inline cpp/src/gandiva/tests/projector_test.cc:212]: can remove the commented code

File: cpp/src/gandiva/engine.h (hunk 16 above)
Comment (projjal) [reply cpp/src/gandiva/engine.h:61]: Ok. looks like earlier versions didnot contain this method. Also looking at setObjectCache impl it just sets the member variable so no need to check for error

File: cpp/src/gandiva/llvm_generator.h (hunk 3 above)
Comment (augustoasilva) [reply cpp/src/gandiva/llvm_generator.h:64]: Yes, I safely removed the duplicated code.

File: cpp/src/gandiva/engine.h (hunk 16 above)
Comment (augustoasilva) [reply cpp/src/gandiva/engine.h:61]: Ok, so I will just leave it withouht the error checking for compability with older versions

File: cpp/src/gandiva/tests/projector_test.cc (hunk 17 above)
Comment (augustoasilva) [reply cpp/src/gandiva/tests/projector_test.cc:212]: Removed it

File: cpp/src/gandiva/greedy_dual_size_cache.h [hunk 18]
```diff
@@ -133,6 +138,14 @@ class GreedyDualSizeCache {
     priority_set_.clear();
   }
 
+  size_t GetCacheSize() { return cache_size_; }
+
+  std::string ToString() {
```
[the remark this answers is not in the document]
Comment (augustoasilva) [reply cpp/src/gandiva/greedy_dual_size_cache.h:143]: Not required, only used for debugging. Should remove it and also remove the lines that call it?

File: cpp/src/gandiva/llvm_generator.h (hunk 4 above)
Comment (augustoasilva) [reply cpp/src/gandiva/llvm_generator.h]: Sorry, it seemed right. But I fixed with your suggestion.

File: cpp/src/gandiva/greedy_dual_size_cache.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (projjal) [reply cpp/src/gandiva/greedy_dual_size_cache.h:143]: I see its called after buildProject/buildFilter. Its not really required. Lets remove it.

File: cpp/src/gandiva/greedy_dual_size_cache.h [hunk 19]
```diff
@@ -17,6 +17,8 @@
 
 #pragma once
 
+#include <llvm/ADT/SmallString.h>
```
Comment (projjal) [inline cpp/src/gandiva/greedy_dual_size_cache.h:20]: this is not required

Comment (augustoasilva) [reply cpp/src/gandiva/greedy_dual_size_cache.h:20]: Ok, I've used the "Optmize imports" from Clion and it did not remove it before, so I removed it.

File: cpp/src/gandiva/base_cache_key.h [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+      return false;
+    }
+
+    return true;
+  }
+
+  bool operator!=(const BaseCacheKey& other) const { return !(*this == other); }
+
+ private:
+  uint64_t hash_code_;
```
Comment (projjal) [inline cpp/src/gandiva/base_cache_key.h:80]: > size_t

File: cpp/src/gandiva/cache.cc [hunk 21]
```diff
@@ -20,10 +20,10 @@
...
-static const int DEFAULT_CACHE_SIZE = 500;
+static const size_t DEFAULT_CACHE_SIZE = 128 * 1024 * 1024;  // 256 MiB
 
-int GetCapacity() {
-  int capacity;
+size_t GetCapacity() {
+  size_t capacity;
   const char* env_cache_size = std::getenv("GANDIVA_CACHE_SIZE");
   if (env_cache_size != nullptr) {
     capacity = std::atoi(env_cache_size);
```
Comment (projjal) [inline cpp/src/gandiva/cache.cc:29]: since it is calculating size it will be capped to int_max (2GB). Better to parse it as long.

File: cpp/src/gandiva/cache.cc [hunk 22]
```diff
@@ -20,10 +20,10 @@
...
 namespace gandiva {
 
-static const int DEFAULT_CACHE_SIZE = 500;
+static const size_t DEFAULT_CACHE_SIZE = 128 * 1024 * 1024;  // 256 MiB
 
-int GetCapacity() {
-  int capacity;
+size_t GetCapacity() {
+  size_t capacity;
   const char* env_cache_size = std::getenv("GANDIVA_CACHE_SIZE");
```
Comment (projjal) [inline cpp/src/gandiva/cache.cc:27]: rename the constant as GANDIVA_CACHE_SIZE_BYTES (otherwise it will pick up the earlier set value on existing machines)

File: cpp/src/gandiva/base_cache_key.h (hunk 20 above)
Comment (augustoasilva) [reply cpp/src/gandiva/base_cache_key.h:80]: Changed it to size_t type

File: cpp/src/gandiva/engine.h (hunk 16 above)
Comment (projjal) [inline cpp/src/gandiva/engine.h:61]: status is not required here. just make it void.

File: cpp/src/gandiva/filter.cc [hunk 23]
```diff
@@ -78,6 +79,10 @@ std::string FilterCacheKey::ToString() const {
   return ss.str();
 }
 
+std::string FilterCacheKey::GetExpressionsAsString() const {
```
Comment (projjal) [inline cpp/src/gandiva/filter.cc:82]: nit: GetExpressionAsString?

File: cpp/src/gandiva/cache.cc (hunk 21 above)
Comment (augustoasilva) [reply cpp/src/gandiva/cache.cc:29]: Done

File: cpp/src/gandiva/cache.cc (hunk 22 above)
Comment (augustoasilva) [reply cpp/src/gandiva/cache.cc:27]: Makes sense, changed to it

File: cpp/src/gandiva/filter.cc (hunk 23 above)
Comment (augustoasilva) [reply cpp/src/gandiva/filter.cc:82]: This function is to be used only in the base_cache_key so it also stores the expressions and compare them when comparing one base_cache_key against another.

File: cpp/src/gandiva/engine.h (hunk 16 above)
Comment (augustoasilva) [reply cpp/src/gandiva/engine.h:61]: Done

File: cpp/src/gandiva/filter.cc (hunk 23 above)
Comment (projjal) [reply cpp/src/gandiva/filter.cc:82]: yes I understood that. I meant the name here should be getExpression instead of getExpressions

File: cpp/src/gandiva/projector.h [hunk 24]
```diff
@@ -34,6 +34,34 @@ namespace gandiva {
 
 class LLVMGenerator;
 
+class ProjectorCacheKey {
```
Comment (projjal) [inline cpp/src/gandiva/projector.h:37]: ProjectorCacheKey and FilterCacheKey are not really required now since we are now using a single cache and you are anyway creating basecachekey from them. Just create a single class named ExpressionCacheKey

Comment (augustoasilva) [reply cpp/src/gandiva/projector.h:37]: Done, now there is only the ExpressionCacheKey

File: cpp/src/gandiva/filter.cc (hunk 23 above)
Comment (augustoasilva) [reply cpp/src/gandiva/filter.cc]: There is no need for it anymore, as it were only used on the ExpressionCacheKey and now the ExpressionCacheKey is already built with the expressions from projector or filter, not the old ProjectorCacheKey or FilterCacheKey

File: cpp/src/gandiva/cache.h [hunk 25]
```diff
@@ -39,17 +39,17 @@ class Cache {
...
+  ValueType GetObjectCode(KeyType cache_key) {
     arrow::util::optional<ValueCacheObject<ValueType>> result;
     mtx_.lock();
     result = cache_.get(cache_key);
     mtx_.unlock();
     return result != arrow::util::nullopt ? (*result).module : nullptr;
   }
 
-  void PutModule(KeyType cache_key, ValueCacheObject<ValueType> valueCacheObject) {
+  void PutObjectCode(KeyType& cache_key, ValueCacheObject<ValueType> object_code) {
```
Comment (projjal) [inline cpp/src/gandiva/cache.h:50]: pass the cacheobject also as reference

Comment (augustoasilva) [reply cpp/src/gandiva/cache.h:50]: Done

**Review by projjal — APPROVED** [review]

**Review by recardozo — COMMENTED** [review]
The best pull request <3
